File: skid_steer_ws/src/skid_steer_controller/src/simple_controller.cpp
#include "skid_steer_controller/simple_controller.hpp"
#include <Eigen/Geometry>

using std::placeholders::_1;


SimpleController::SimpleController(const std::string & name)
    : Node(name){
        declare_parameter("wheel_radius", 0.059);
        declare_parameter("wheel_separation", 0.55645);

        wheel_radius = get_parameter("wheel_radius").as_double();
        wheel_separation = get_parameter("wheel_separation").as_double();

        RCLCPP_INFO_STREAM(get_logger(), "Using wheel_radius " << wheel_radius);
        RCLCPP_INFO_STREAM(get_logger(), "Using wheel_separation " << wheel_separation);
        wheel_cmd_publisher = create_publisher<std_msgs::msg::Float64MultiArray>("/simple_velocity_controller/commands", 10);
        vel_subscriber = create_subscription<geometry_msgs::msg::TwistStamped>("/skid_steer_controller/cmd_vel", 10, 
                        std::bind(&SimpleController::velCallback, this, _1));
    
        speed_conversion << wheel_radius/2,                         wheel_radius/2, 
                            wheel_radius/wheel_separation, -wheel_radius/wheel_separation;

        RCLCPP_INFO_STREAM(get_logger(), "The conversion matrix is \n" << speed_conversion);

    };

void SimpleController::velCallback(const geometry_msgs::msg::TwistStamped & msg){
    Eigen::Vector2d robot_speed(msg.twist.linear.x, msg.twist.angular.z); 
    Eigen::Vector2d wheel_speed = speed_conversion.transpose() * robot_speed;

    std_msgs::msg::Float64MultiArray wheel_speed_msg;
    wheel_speed_msg.data.push_back(wheel_speed.coeff(1));
    wheel_speed_msg.data.push_back(wheel_speed.coeff(0));
    wheel_speed_msg.data.push_back(wheel_speed.coeff(1));
    wheel_speed_msg.data.push_back(wheel_speed.coeff(0));
    wheel_speed_msg.data.push_back(wheel_speed.coeff(1));
    wheel_speed_msg.data.push_back(wheel_speed.coeff(0));
    wheel_cmd_publisher->publish(wheel_speed_msg);
};

int main(int argc ,char* argv[]){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleController>("simple_controller");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}                       

File: skid_steer_ws/src/skid_steer_controller/skid_steer_controller/simple_controller.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
import rclpy.node
from std_msgs.msg import Float64MultiArray
from geometry_msgs.msg import TwistStamped
import numpy as np

class SimpleController(Node):
    def __init__(self):
        super().__init__("simple_controller")

        self.declare_parameter("wheel_radius", 0.059)
        self.declare_parameter("wheel_separation", 0.55645)

        self.wheel_radius = self.get_parameter("wheel_radius").get_parameter_value().double_value
        self.wheel_separation = self.get_parameter("wheel_separation").get_parameter_value().double_value

        self.get_logger().info("Using wheel_radius %f" % self.wheel_radius)
        self.get_logger().info("Using wheel_separation %f" % self.wheel_separation)

        self.wheel_cmd_publisher = self.create_publisher(Float64MultiArray, "simple_velocity_controller/commands", 10)
        self.vel_subscriber = self.create_subscription(TwistStamped, "skid_steer_controller/cmd_vel", self.velCallback, 10)
        self.speed_conversion = np.array([[self.wheel_radius/2, self.wheel_radius/2], 
                                          [self.wheel_radius/self.wheel_separation, -self.wheel_radius/self.wheel_separation]])
        self.get_logger().info("The conversion matrix is %s" % self.speed_conversion)
    def velCallback(self, msg):
        robot_speed = np.array([[msg.twist.linear.x],
                                [msg.twist.angular.z]])
        wheel_speed = np.matmul(np.linalg.inv(self.speed_conversion), robot_speed)
        wheel_speed_msg = Float64MultiArray()
        wheel_speed_msg.data=[wheel_speed[1, 0], wheel_speed[0, 0]]
        self.wheel_cmd_publisher.publish(wheel_speed_msg)

def main():
    rclpy.init()
    simple_controller = SimpleController()
    rclpy.spin(simple_controller)
    simple_controller.destroy_node()
    rclpy.shutdown()
    
if __name__ == '__main__':
    main()